idea-plugin/src/main/kotlin/arrow/meta/dsl/ide/editor/lineMarker/LineMarkerSyntax.kt

```diff
@@ -17,38 +18,42 @@ interface LineMarkerSyntax {
 
   /**
    * This technique adds an LineMarker on the specified PsiElement similar to the Recursive Kotlin Icon [org.jetbrains.kotlin.idea.highlighter.KotlinRecursiveCallLineMarkerProvider]
-   * or Suspended Icon [org.jetbrains.kotlin.idea.highlighter.KotlinSuspendCallLineMarkerProvider]
+   * or Suspended Icon [org.jetbrains.kotlin.idea.highlighter.KotlinSuspendCallLineMarkerProvider].
    * TODO: Add more Techniques such as the one from Elm
    */
   @Suppress("UNCHECKED_CAST")
   fun <A : PsiElement> IdeMetaPlugin.addLineMarkerProvider(
     icon: Icon,
-    matchOn: (PsiElement) -> A?,
+    transform: (PsiElement) -> A?,
     message: (element: A) -> String = Noop.string1(),
     placed: GutterIconRenderer.Alignment = GutterIconRenderer.Alignment.RIGHT
   ): ExtensionPhase =
     addLineMarkerProvider(
-      matchOn,
-      { a ->
-        lineMarkerInfo(icon, a, message as (PsiElement) -> String, placed)
+      transform,
+      {
+        it.children.firstOrDefault(it).let { p -> lineMarkerInfo(icon, p, message as (PsiElement) -> String, placed) }
```

Comment:
Why is `firstOrDefault()` called here? If this is to get a leaf element, then it's not correct. 
A `PsiElement` is either a leaf (i.e. no child elements) or a composite element with `PsiElements` as child elements. Each of these children can now be leads or composite elements.

Here's the subtree of a `KtClass` (Tools->View Psi structure of current file...)
![Screenshot_20191026_172031](https://user-images.githubusercontent.com/11473063/67621933-efef9100-f814-11e9-8c35-f5989da5d7bd.png)

The identifier is a leaf, but not necessarily the first. And there will be other provider implementations which don't want to limit the marker for the first child.

And, in general, a line marker provider should always add the marker to the `PsiElement` passed in by IntelliJ and not to a child element. IntelliJ keeps a cache of markers for `PsiElements` and you'll probably see unexpected behaviour with this implementation.
I'll add a general note, too.

Reply:
I agree with a few points of you, but the emphasis for this implementation is not to get the PsiIdentifier, because not all PsiElements, users will match on have that Property.

Thus we can simplify IntelliJ's suggestion to any leave element that a user passes. 
BTW the behavior stays the same. But we can verify that later with tests.
I am open for suggestions on how to get the deepest leave. I am trying out different things

Reply:
Could you explain what you mean by `the emphasis for this implementation is not to get the PsiIdentifier, because not all PsiElements, users will match on have that Property.`? My understanding is that the PsiIdentifier will be used for higherkind line marker, for example. Other line marker providers would annotate other leaf element.

What I mean is that the caller of `addLineMarkerProvider` has to traverse the PSI tree [here](https://github.com/arrow-kt/arrow-meta/blob/7bb2d0ccdb91928312c1d7f7b2f3bd9e53810048/idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/plugins/higherkinds/HigherKindsIdePlugin.kt#L23). My other PR has an example [here](https://github.com/arrow-kt/arrow-meta/blob/7cf6a6f22913853dd889f79f3da4de1e498bdeb7/idea-plugin/src/main/kotlin/arrow/meta/plugin/idea/plugins/higherkinds/HigherKindsIdePlugin.kt#L19).

With the current implementation you're getting this:
![Screenshot_20191026_185325](https://user-images.githubusercontent.com/11473063/67623127-37c8e500-f822-11e9-974c-c64a9372d4f8.png)
Here `@kotlin.Deprecated` is annotated. That's probably not correct (I assume) and `@kotlin.Deprecated` isn't a leaf element, too.

My understanding is that you'd like to put the line marker next to the class name.

To get the leaf element of a PsiElement you could use `PsiTreeUtil.getDeepestFirst()`, but that will only work as expected with a Psi structure where each parent contains only one child.
For the first example in the screenshot this would return the token `@` and the token `/**` for the 2nd example.

Reply:
Yes I am aware of that I thought you meant with behavior, that it doesn't subscribe to the right component. 
Makes sense. But the overall goal is not to pass a PsiIdentifier in `transform`. 
As you do in your PR. 

I am currently implementing something which looks for the deepest leave.

Reply:
(Rephrased my comments above, apologies if it sounded too direct or unfriendly.)
`PsiTreeUtil.getDeepestFirst` will return the leaf at the leftmost position in a PsiElement, `getDeepestLast` will return the rightmost element (referring to the position in the editor).
`getDeepestFirst` might be what's needed here for the annotation.

Reply:
Thank you I will try both of them out. 
No worries. It happens to me accidentally, too. May it's whenever we think in German and speak in English  :slightly_smiling_face:

Reply:
Maybe we could do somthing like: 
Give me the PsiIdentifier if that doesnt exist subscribe to the `Annotation`.
Would that satisfy the IntelliJ suggestion @jansorg ?

Reply:
The result is in the latest push.